### ISSUE TITLE
Add TripletLearningAlgorithm to DL4J

### ISSUE THREAD

**Issue by crockpotveggies** [deeplearning4j/deeplearning4j#2216]
Opening this issue to centralize discussion for adding a `TripletLearningAlgorithm` to Deeplearning4j. Triplet Embeddings have been shown to be extremely useful for calculating euclidean distance for classification and clustering, and personally solves a lot of business problems we have at Bernie AI.

I'd like to implement Triplet Embedding in DL4J and generalize it to `MultiLayerNetwork` and `ComputationGraph` so that it can also be shared with future users. The purpose of this issue is to consult with authors to understand the architecture of what I'm implementing, and ensure that it is planned correctly. As per Gitter chat, sounds like @raver119 and @AlexDBlack are the two helpful people to provide input.

![image](https://cloud.githubusercontent.com/assets/1445295/19708729/a1a4ce38-9ad5-11e6-98c3-562f077b730a.png)

Rough strategy for implementation:
1. Create a `TripletLearningAlgorithm` in `org.deeplearning4j.models.embeddings.learning`
2. Base some of the work using embeddings via @raver119 's `ElementsLearningAlgorithm`
3. Write a `TripletLoss` function in ND4J
4. Implement triplet generation in DataVec so datasets are automatically preprocessed (maybe this should be called `org.datavec.common.TripletConverter`?)
5. What else is needed if we want this to generalize to MLN and CG?

Related literature using Triplet Embeddings:
[FaceNet](https://arxiv.org/pdf/1503.03832v3.pdf)
[OpenFace](http://reports-archive.adm.cs.cmu.edu/anon/2016/CMU-CS-16-118.pdf)

Similar implementations include:
[Torch's TripletEmbedding](https://github.com/Atcold/torch-TripletEmbedding/blob/742599ce761b2d6cf128068b8f6fd327ad633a31/TripletEmbedding.lua) and [an example](https://github.com/Atcold/torch-TripletEmbedding/blob/742599ce761b2d6cf128068b8f6fd327ad633a31/xmp/fresh-embedding.lua)
[Caffe TripletLossLayer](https://github.com/deepxkn/triplet/blob/master/triplet/tripletloss_layer.py)

Your thoughts and ideas here will help me implement this, thanks!

**Comment by AlexDBlack** [comment]
OK, so I've been reading this paper: https://arxiv.org/pdf/1412.6622v3.pdf

So, as I understand it:
- 3 inputs: x-, x, x+ where:
    - x and x+ are of the same class
    - x and x- are of different class
- Pass each input through the same network
    - Output of the network (3rd last layer) is your embedding, per example
- second last layer: output is 2 real-valued numbers
    - l2Norm(net(x)-net(x-)) and l2Norm(net(x)-net(x+))
- last layer: softmax on the l2 norm values
- train the network predict which input pairs (x,x+) and (x,x-) are of the same class



So, implementing this: I see the following issues
- Network has 3 inputs: means we can use ComputationGraph only I think
- Shared parameters
    - Not currently supported by DL4J
    - But: a simple/clean solution here: stack the examples. So with 3x [nExamples, inputSize] inputs -> stack to a single [3x nExamples, inputSize] vector, pass through the network, then unstack before the L2 layer
    - That way: only ever have 1 copy of the network. And implementing it is just a couple of relatively simple custom graph vertexes for the stacking/unstacking
- For data, RecordReaderMultiDataSetIterator should probably provide what we need. Depending on source data format
- Do we need a custom loss function? Maybe, maybe not...
    - If we add a L2 norm layer (or probably, an l2 norm GraphVertex given it has multiple inputs): then it's just a standard softmax layer on top of that... no custom loss function required.
    - But we could combine the last 2 layers into 1 (as far as implementation is concerned): then this could be done as a custom loss function I think... (I think I'd prefer the l2 norm vertex approach though - seems cleaner to me)
- I'd like to have this be general-purpose: work for CNNs, MLPs, embedding layer etc for the network (or even RNNs, via last time step vertex functionality). I don't see any of that being a problem here.

**Comment by AlexDBlack** [comment]
Actually: hm. Re: the loss function: an OutputLayer + softmax on top of the L2 norm values wouldn't be exactly the same as the loss function there (output layer has weights/biases). But adding an output layer that does just softmax (or whatever) without the MLP part would be easy.

**Comment by crockpotveggies** [comment]
Okay to reiterate what you've said and to ensure I understand you correctly:

- We don't need a `TripletLearningAlgorithm` but we will need some sort of custom layer 
- InputLayer is 3x [nExamples, inputSize]  which is stacked and unstacked (I assume I'll be using something like `MergeVertex`?
- Last layer is a softmax on the L2 norm values

Some concerns:

- Since I'm still getting familiarized with ComputationGraph, what layer should I be using to unstack the triplets?
- Regarding softmax output layer, I'm correct in thinking this is a custom layer? Is this just `TripletLossLayer`? (should I be implementing both `TripletLoss` and `TripletLossLayer`?)

**Comment by AlexDBlack** [comment]
Right, I think we should just build thing out of ComputationGraph components - no separate model type or anything necessary. That way we get everything else (spark, model serialization, gradient checks etc) for free.

Stacking would be very similar to MergeVertex - just combining on different dimension. Would be general purpose, not just for triplet network.

For stacking/unstacking examples: we'd need to implement a custom GraphVertex for that. Nothing we currently have would do that. Wouldn't be hard to add.

For softmax output: I'm thinking we make a version of OutputLayer minus the MLP part. That has the restriction that nIn = nOut for that one. Again, worth having in general.

**Comment by crockpotveggies** [comment]
Some quick clarification on `That has the restriction that nIn = nOut for that one`. Are you saying it needs the restriction? Or that we *don't* want the restriction?

I'm assuming the latter.

**Comment by AlexDBlack** [comment]
No, I'm saying that it's like ActivationLayer - it wouldn't be a layer in the real sense (doesn't have parameters) but just applies the activation and loss functions. Because of that, nIn=nOut is a strict requirement (same as activation layer).

**Comment by crockpotveggies** [comment]
Resolved by https://github.com/deeplearning4j/deeplearning4j/pull/2275

**Comment by bocse** [comment]
How can the network be used after training with the Triplet Loss Function?
The current structure of the network requires 3 images as input. How can the network then be used (in evaluation) with just one input?

Thank you.

**Comment by crockpotveggies** [comment]
It's up to you to develop the transfer learning strategy, but strategies include copying parameters and snipping the network.

**Comment by agibsonccc** [comment]
Of note here is we are building an explicit transfer learning api as well.

**Comment by bocse** [comment]
@agibsonccc i'd very much like to try it out. Could you maybe point me in the right direction? :)

**Comment by agibsonccc** [comment]
https://github.com/deeplearning4j/dl4j-examples/tree/105ea7b9b9625cc8138d89f10b470d0b41680c6b/dl4j-examples/src/main/java/org/deeplearning4j/examples/transferlearning

Our examples contain *everything* look here first before asking questions.
